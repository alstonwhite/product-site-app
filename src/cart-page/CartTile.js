import React from 'react';
import { Link } from "react-router-dom";

import logo from '../logo.svg';

const CartTile = ({product, onUpdate, onRemove}) => {
    
    return (
      <div className="cart-tile" id="cart-tile">
        <Link className="cart-tile__content" to={`/product/${product.id}`}>  
          <img alt='logo' className='cart-tile__content__img' src={product.image1 ? product.image1.fields.file.url : logo}/>
          <div className="cart-tile__content__name">{product.title}</div>  
          <div className="cart-tile__content__price">
              <span>$</span>
              {product.price}
            </div>
        </Link>
        <div className="cart-tile__action">
            {/* check input (positive integers only, what if deleted?) */}
            <input 
                className="cart-tile__action__qty"
                type="number"
                defaultValue={product.quantity}
                onChange={event=>onUpdate(product,event.target.value)}
            />
            <button
                className="cart-tile__action__remove"
                onClick={()=>onRemove(product)}
            >
            X
            </button>
        </div>
      </div>
    );
  };
  
  export default CartTile;