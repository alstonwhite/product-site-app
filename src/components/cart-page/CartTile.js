import React from 'react';
import { Link } from "react-router-dom";

import { useDispatch } from "react-redux";
import { upDateQtyCartR, removeItemCartR } from '../../redux/actions';

import logo from '../../assets/particle-logo2.jpeg';

const CartTile = ({product}) => {

  const dispatch = useDispatch();
  const onUpdate = (product, qty) => dispatch(upDateQtyCartR(product, qty));
  const onRemove = product => dispatch(removeItemCartR(product));

  const validateInput = (input)  => {
    if (input <=0 ) {
      onRemove(product);
    } else {
      onUpdate(product, input);
    }
  }

  // note: src={product.image1 ? product.image1.fields.file.url : logo} renders 'logo' as placeholder if Contentful product does not have image

  return (
    <div className="cart-tile" id="cart-tile" data-testid="cart-tile">
      <Link className="cart-tile__content" to={`/product/${product.id}`}>  
        <img alt='logo' className='cart-tile__content__img' src={product.image1 ? product.image1.fields.file.url : logo}/>
        <div className="cart-tile__content__name">{product.title}</div>  
        <div className="cart-tile__content__price">
            <span>$</span>
            {product.price}
          </div>
      </Link>
      <div className="cart-tile__action">
          <input 
              className="cart-tile__action__qty"
              type="number"
              min="0"
              step="1"
              defaultValue={product.quantity}
              onBlur={e => validateInput(e.target.value)}
          />
          <button
              className="cart-tile__action__remove btn"
              onClick={()=>onRemove(product)}
          >
          X
          </button>
      </div>
    </div>
  );
};
  
export default CartTile;