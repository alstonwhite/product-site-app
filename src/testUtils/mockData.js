export const testSections = 
    [
        {
            id: "0",
            slug: "section1",
            sectionTitle: "Section 1",
            active: true
        },
        {
            id: "1",
            slug: "section2",
            sectionTitle: "Section 2",
            active: true
        }
    ]

export const testProducts = 
    [
        {
            id: "0",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 1",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "1",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 2",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "2",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 3",
            description: "Description",
            productSection: {fields: {slug: "section2"}},
            active: true
        }
    ]

export const testProduct = 
    {
        id: 0,
        image1: {fields: {file: {url: "1"}}},
        image2: {fields: {file: {url: "2"}}},
        image3: {fields: {file: {url: "3"}}},
        price: 5,
        title: "Product 1",
        description: "Description",
        productSection: {fields: {slug: "section1"}},
        active: true
    }

export const testCartProduct = 
[{
    id: 0,
    image1: {fields: {file: {url: "1"}}},
    image2: {fields: {file: {url: "2"}}},
    image3: {fields: {file: {url: "3"}}},
    price: 5,
    title: "Product 1",
    description: "Description",
    productSection: {fields: {slug: "section1"}},
    active: true,
    quantity: 1
}]
