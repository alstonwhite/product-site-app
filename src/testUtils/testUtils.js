import React from "react";
import { Router, Route } from "react-router-dom";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";

export const renderWithRouter = (ui, 
    {
        path="/",
        route="/",
        history = createMemoryHistory({ initialEntries: [route] }) 
    }={}) => ( 
    render( 
        <Router history={history}> 
            <Route path={path} children={ui} /> 
        </Router> 
    ) 
)