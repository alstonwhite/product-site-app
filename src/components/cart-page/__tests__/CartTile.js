import React from 'react'
import { cleanup, fireEvent, screen } from '@testing-library/react'
import CartTile from '../CartTile'
import { testCartProduct } from '../../../testUtils/mockData'
import { renderWithRouter } from '../../../testUtils/testUtils'


afterEach(cleanup)

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
    useDispatch: () => mockDispatch
}));

test('calls onUpdate with the product and value when input updated with positive int', () => {
    renderWithRouter(<CartTile product={testCartProduct[0]}/>, {path: "/cart", route: "/cart"})

    screen.getByDisplayValue(1).focus();
    fireEvent.change(screen.getByDisplayValue(1), {target: {value: '2'}})
    screen.getByText('X').focus();

    expect(mockDispatch).toHaveBeenCalledTimes(1)
    expect(mockDispatch).toHaveBeenCalledWith({
        type: 'CART_UPDATE',
        payload: {
            product: testCartProduct[0],
            quantity: '2'
        }
    })
})

test('calls onRemove with the product and value when input updated with 0', () => {
    renderWithRouter(<CartTile product={testCartProduct[0]}/>, {path: "/cart", route: "/cart"})
  
    screen.getByDisplayValue(1).focus();
    fireEvent.change(screen.getByDisplayValue(1), {target: {value: '0'}})
    screen.getByText('X').focus();

    expect(mockDispatch).toHaveBeenCalledTimes(1)
    expect(mockDispatch).toHaveBeenCalledWith({
        type: 'CART_REMOVE',
        payload: testCartProduct[0]
    })
})

test('calls onRemove with the product when button clicked', () => {
    renderWithRouter(<CartTile product={testCartProduct[0]}/>, {path: "/cart", route: "/cart"})

    fireEvent.click(screen.getByText('X'))
  
    expect(mockDispatch).toHaveBeenCalledTimes(1)
    expect(mockDispatch).toHaveBeenCalledWith({
        type: 'CART_REMOVE',
        payload: testCartProduct[0]
    })
})
  
test('snapshot', () => {
    const {container} = renderWithRouter(<CartTile product={testCartProduct[0]}/>, {path: "/cart", route: "/cart"})
    expect(container.firstChild).toMatchSnapshot()
})