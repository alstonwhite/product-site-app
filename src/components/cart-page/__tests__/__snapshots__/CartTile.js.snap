// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="cart-tile"
  data-testid="cart-tile"
  id="cart-tile"
>
  <a
    class="cart-tile__content"
    href="/product/0"
  >
    <img
      alt="logo"
      class="cart-tile__content__img"
      src="1"
    />
    <div
      class="cart-tile__content__name"
    >
      Product 1
    </div>
    <div
      class="cart-tile__content__price"
    >
      <span>
        $
      </span>
      5
    </div>
  </a>
  <div
    class="cart-tile__action"
  >
    <input
      class="cart-tile__action__qty"
      min="0"
      step="1"
      type="number"
      value="1"
    />
    <button
      class="cart-tile__action__remove btn"
    >
      X
    </button>
  </div>
</div>
`;
